feat(examples) Add `LeRobot` quickstart-example (#4607)
adap/flower: examples/quickstart-lerobot/README.md

---

> tags: [robotics, ai, zk0]

Reviewer: ```suggestion
tags: [quickstart, vision, robotics]
```

dataset: [pusht]
framework: [lerobot]
---

> # Federated Learning with HuggingFace LeRobot and Flower (Quickstart Example)

Reviewer: ```suggestion
# Federated Learning with Hugging Face LeRobot and Flower (Quickstart Example)
```

In this example, we will federate the training of a [Diffusion](https://arxiv.org/abs/2303.04137) policy on the [PushT](https://huggingface.co/datasets/lerobot/pusht/tree/v1.3) dataset. The data will be downloaded and partitioned using [Flower Datasets](https://flower.ai/docs/datasets/). This example runs best when a GPU is available.

![](_static/render_compose.gif)

## Set up the project

### Clone the project

Start by cloning the example project. We prepared a single-line command that you can copy into your shell which will checkout the example for you:

```shell
git clone --depth=1 https://github.com/adap/flower.git _tmp \
		&& mv _tmp/examples/quickstart-lerobot . \
		&& rm -rf _tmp && cd quickstart-lerobot
```

This will create a new directory called `quickstart-lerobot` containing the following files:

```shell
quickstart-lerobot
├── lerobot_example
│   ├── __init__.py
│   ├── client_app.py   # Defines your ClientApp
│   ├── server_app.py   # Defines your ServerApp
│   ├── task.py         # Defines your model, training and data loading
│   ├── lerobot_federated_dataset.py   # Defines the dataset

> |   └── configs/		# configuration files
> │ 		├── env/        	# gym environment config
> |   	├── policy/			# policy config
> |   	└── default.yaml 	# default config settings
> |

Reviewer: ```suggestion
│   └── configs/		# configuration files
│ 		├── env/        	# gym environment config
│   	├── policy/			# policy config
│   	└── default.yaml 	# default config settings
│
```

Tiny edits for consistency.

```bash
# Run with the default federation (CPU only)
flwr run .
```

Run the project in the `local-simulation-gpu` federation that gives CPU and GPU resources to each `ClientApp`. By default, at most 2x`ClientApp` (using ~2 GB of VRAM each) will run in parallel in each available GPU. Note you can adjust the degree of parallelism but modifying the `client-resources` specification. Running with the settings as in the `pyproject.toml` it takes 1h in a 2x RTX 3090 machine.

```bash
# Run with the `local-simulation-gpu` federation
flwr run . local-simulation-gpu
```

You can also override some of the settings for your `ClientApp` and `ServerApp` defined in `pyproject.toml`. For example

```bash
flwr run . local-simulation-gpu --run-config "num-server-rounds=5 fraction-fit=0.1"
```

### Result output

Results of training steps for each client and server logs will be under the `outputs/` directory. For each run there will be a subdirectory corresponding to the date and time of the run. For example:

```shell
outputs/date_time/
├── evaluate  # Each subdirectory contains .mp4 renders generated by clients
│   ├── round_5	# Evaluations in a given round

> |	|   ├── client_3
> |	|	...	└── rollout_20241207-105418.mp4 # render .mp4 for client at a given round
> |	|	└── client_1
> |   ...

Reviewer: ```suggestion
│	│   ├── client_3
│	│	...	└── rollout_20241207-105418.mp4 # render .mp4 for client at a given round
│	│	└── client_1
│   ...
```

Trivial edits for consistency.

---

# Federated Learning with HuggingFace LeRobot and Flower (Quickstart Example)

This is an introductory example to using [🤗LeRobot](https://huggingface.co/lerobot) with [🌼Flower](https://flower.ai/). It demonstrates that it is feasible to collaboratively train a robotics AI model in remote environments with their local data and then aggregated it in a shared model.

In this example, we will federate the training of a [Diffusion](https://arxiv.org/abs/2303.04137) policy on the [PushT](https://huggingface.co/datasets/lerobot/pusht/tree/v1.3) dataset. The data will be downloaded and partitioned using [Flower Datasets](https://flower.ai/docs/datasets/). This example runs best when a GPU is available.

## Set up the project

### Clone the project

Start by cloning the example project. We prepared a single-line command that you can copy into your shell which will checkout the example for you:

```shell
git clone --depth=1 https://github.com/adap/flower.git _tmp \
		&& mv _tmp/examples/quickstart-lerobot . \
		&& rm -rf _tmp && cd quickstart-lerobot
```

This will create a new directory called `quickstart-lerobot` containing the following files:

```shell
quickstart-lerobot

> ├── lerobot_example

Reviewer 2: directories `configs` and `policy` are missing

Author: I force added them now. Not sure why git excluded earlier. Can you see them?

Reviewer 2: perfect. I see them now 💯
